src/main/java/frc/robot/subsystems/shooter/Shooter.java

```diff
@@ -0,0 +1,160 @@
+        main.setSensorPhase(Ports.Shooter.IS_SENSOR_INVERTED);
+
+        main.setNeutralMode(NeutralMode.Coast);
+        aux.setNeutralMode(NeutralMode.Coast);
+
+        main.configForwardLimitSwitchSource(LimitSwitchSource.Deactivated, LimitSwitchNormal.Disabled);
+        main.configReverseLimitSwitchSource(LimitSwitchSource.Deactivated, LimitSwitchNormal.Disabled);
+
+        main.enableVoltageCompensation(true);
+        aux.enableVoltageCompensation(true);
+
+        main.configVoltageCompSaturation(Constants.NOMINAL_VOLTAGE);
+        aux.configVoltageCompSaturation(Constants.NOMINAL_VOLTAGE);
+
+        aux.follow(main);
+
+        this.stateSpacePredictor = constructLinearSystem();
+        this.velocityEstimator = new LinearRegression(readCSV());
+    }
+
+    /**
+     * Initialize the linear system to the default values in order to use the state-space.
+     *
+     * @return an object that represents the model to reach the velocity at the best rate.
+     */
+    private LinearSystemLoop<N1, N1, N1> constructLinearSystem() {
+        // https://file.tavsys.net/control/controls-engineering-in-frc.pdf Page 76
+        Vector<N1> A = VecBuilder.fill(-Math.pow(GEAR_RATIO, 2) * kT / (kV * OMEGA * J));
+        Vector<N1> B = VecBuilder.fill(GEAR_RATIO * kT / (OMEGA * J));
+        LinearSystem<N1, N1, N1> stateSpace = new LinearSystem<>(A, B, Matrix.eye(Nat.N1()), new Matrix<>(Nat.N1(), Nat.N1()));
+        KalmanFilter<N1, N1, N1> kalman = new KalmanFilter<>(Nat.N1(), Nat.N1(), stateSpace,
+                VecBuilder.fill(MODEL_TOLERANCE),
+                VecBuilder.fill(ENCODER_TOLERANCE),
+                Constants.LOOP_PERIOD
+        );
+        LinearQuadraticRegulator<N1, N1, N1> lqr = new LinearQuadraticRegulator<>(stateSpace, VecBuilder.fill(VELOCITY_TOLERANCE),
+                VecBuilder.fill(Constants.NOMINAL_VOLTAGE),
+                Constants.LOOP_PERIOD // time between loops, DON'T CHANGE
+        );
+
+        return new LinearSystemLoop<>(stateSpace, lqr, kalman, Constants.NOMINAL_VOLTAGE, Constants.LOOP_PERIOD);
+    }
+
+    /**
+     * Internal function to create an input stream reader, in order values from the shooting experiments.
+     *
+     * @return a reader to the CSV.
+     */
+    private InputStreamReader readCSV() {
+        InputStream is = getClass().getResourceAsStream(PATH_TO_CSV);
+        assert is != null : "Can't create input stream";
+
+        return new InputStreamReader(is);
+    }
+
+    /**
+     * @return the velocity of the shooter. [RPS]
+     * @see #setVelocity(double, double)
+     */
+    public double getVelocity() {
+        return unitModel.toVelocity(main.getSelectedSensorVelocity());
+    }
+
+    /**
+     * Set the velocity to apply by the motor.
+     *
+     * @param velocity the desired velocity at which the motor will rotate. [RPS]
+     * @see #setPower(double)
+     */
+    public void setVelocity(double velocity, double difference) {
```

Comment:
Why not calculate `difference` internally?

Reply:
because it's just make it harder to maintain, I need to handle the case when they first shoot, the case when the shoot after a lot of time, and the usual case... (it's more error prone, they can always use the `LOOP_PERIOD` constant)

Reply:
I don't understand why it will be complicated. Just initialize a "last timestamp" variable in the constructor and update it in `periodic()`.

Reply:
because it's not the interval between every call in the shoot command

Reply:
Why not? You update `lastTime` every loop:
https://github.com/Galaxia5987/robot-2021/blob/82dadb5683073d0f3711a970530c5484db490204/src/main/java/frc/robot/subsystems/shooter/commands/Shoot.java#L35-L37

Reply:
Yes, but the loop interval of the `periodic()` function is different from the loop period of the `execute()` function

Reply:
Don't they both run at 50Hz?

Reply:
idk, but either way, using the current implementation is cleaner, in your way I still need to handle the case that we didn't shoot for a while

Reply:
> but the loop interval of the `periodic()` function is different from the loop period of the `execute()` function

> Don't they both run at 50Hz?

> idk

There is no reason for them to not run at the same frequency.

Reply:
I want to do this in my way, there is nothing wrong about it, even though I think that using timers will break the idea of discretizing the shooter....

Reply:
> I want to do this in my way

Never said you can't.
> I think that using timers will break the idea of discretizing the shooter....

So why are you using a timer?
https://github.com/Galaxia5987/robot-2021/blob/5f17f0282e9630c2105ec8513701d9aaa3106df2/src/main/java/frc/robot/subsystems/shooter/commands/Shoot.java#L18